lib/plug/static.ex

```diff
@@ -67,38 +81,58 @@ defmodule Plug.Static do
     {Plug.Router.Utils.split(at), from, gzip, cache}
   end
 
-  def call(conn = %Conn{method: meth}, {at, from, gzip, cache}) when meth in @allowed_methods do
+  def call(conn = %Conn{method: meth}, {at, from, gzip, cache})
+      when meth in @allowed_methods, do:
     send_static_file(conn, at, from, gzip, cache)
-  end
   def call(conn, _opts), do: conn
 
+  @spec send_static_file(Plug.Conn.t, String.t, Path.t | atom, boolean, boolean)
+    :: Plug.Conn.t
   defp send_static_file(conn, at, from, gzip, cache) do
-    segments = subset(at, conn.path_info)
-    segments = for segment <- List.wrap(segments), do: URI.decode(segment)
+    # subset/2 returns the segments in `conn.path_info` without the segments at
+    # the beginning that are shared with `at`.
+    segments = subset(at, conn.path_info) |> Enum.map(&URI.decode/1)
     path     = path(from, segments)
 
-    cond do
-      segments in [nil, []] ->
+    if invalid_path?(segments) do
+      raise InvalidPathError
+    end
+
+    case file_encoding(conn, path, gzip) do
+      {conn, path} ->
+        if cache, do: conn = set_cache(conn, path)
+
+        content_type = segments |> List.last |> Plug.MIME.path
+
+        conn
+        |> put_resp_header("content-type", content_type)
+        |> send_file(200, path)
+        |> halt
+      :error ->
         conn
-      invalid_path?(segments) ->
-        raise InvalidPathError
-      true ->
-        case file_encoding(conn, path, gzip) do
-          {conn, path} ->
-            if cache do
-              conn = put_resp_header(conn, "cache-control", "public, max-age=31536000")
-            end
-
-            conn
-            |> put_resp_header("content-type", Plug.MIME.path(List.last(segments)))
-            |> send_file(200, path)
-            |> halt
-          :error ->
-            conn
-        end
     end
   end
 
+  @spec set_cache(Plug.Conn.t, Path.t) :: Plug.Conn.t
+  defp set_cache(conn, path)
+  defp set_cache(%Plug.Conn{query_string: ""} = conn, path),
+    do: conn |> put_resp_header("etag", etag_string_for_path(path))
+  defp set_cache(conn, _path),
+    do: conn |> put_resp_header("cache-control", "public, max-age=31536000")
+
+  @spec etag_string_for_path(Path.t) :: String.t
+  defp etag_string_for_path(path) do
+    %File.Stat{size: size, mtime: {{yy, mm, dd}, {h, m, s}}} = File.stat!(path)
+    hash_fun = &:crypto.hash(:md5, &1)
+
+    [size, yy, mm, dd, h, m, s]
+    |> Enum.join("")
+    |> hash_fun.()
+    |> :base64.encode()
```

Comment:
We also won't need to convert anything to string using `Enum.join("")`. For every page served, we need to serve *a lot* of assets, so this should be as fast as possible.